### PR TITLE
chore: use `@web/test-runner` instead of `karma`

### PR DESCRIPTION
**Pull request by SimenB**
## Which problem is this PR solving?

- Attempting to unblock #1152 - i.e. bringing back the browser extension
- Note that this just affects `@opentelemetry/instrumentation-document-load` for now - I wanna get some feedback from the team that that this is something you want before spending more time on it 🙂 

## Short description of the changes

- As mentioned in https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1152#issuecomment-1818373047, `karma` is deprecated, and they [recommend](https://blog.angular.io/moving-angular-cli-to-jest-and-web-test-runner-ef85ef69ceca) using Jest or `@web/test-runner` as a replacement. Seeing as I guess these theses _should_ run in the browser, I elected to go with `@web/test-runner`
- Under the hood `wtr` already uses `mocha`, so that migration is free. However, we also need to replace `webpack` with `rollup`/`esbuild` - that requires some more changes. I was unable to get `assert` to work, so I replaced it with `chai`.
- Coverage is collected, but I'm unsure if CI will correctly pick it up (EDIT: See later comments - the coverage is picked up, but something's weird with the merged report on codecov)

### CODE REVIEW

File: plugins/web/opentelemetry-instrumentation-document-load/test/documentLoad.test.ts
Comment (SimenB) [inline plugins/web/opentelemetry-instrumentation-document-load/test/documentLoad.test.ts:543]: URL is different, and the session ID unique per run. I think this is fine, but happy to change to whatever you want 🙂

File: plugins/web/opentelemetry-instrumentation-document-load/test/documentLoad.test.ts
Comment (SimenB) [inline plugins/web/opentelemetry-instrumentation-document-load/test/documentLoad.test.ts:39]: ideally we'd just get `assert` to work in the browser, but for some reason it gives an error

**Comment by SimenB** [comment]
OK, using Node 16 works with both old webpack and the WTR runner 👍

**Comment by SimenB** [comment]
Hmm, unsure why code coverage was lost - the output file was located: https://github.com/open-telemetry/opentelemetry-js-contrib/actions/runs/6928406995/job/18844148357?pr=1816#step:8:79

Pushed an attempt to also produce json output - let's see 😀

**Comment by SimenB** [comment]
Wait, is it correct? Codecov UI is confusing me 😅 

image

But it is covered if I open `coverage/lcov-report/index.html`

image

No clue 😅

**Comment by SimenB** [comment]
At least running `npm run test:browser && open coverage/lcov-report/index.html` locally shows correct coverage information, so I think the test runner changes are correct. Getting codecov to agree shouldn't be that hard afterwards if it's some sort of permission/from fork weirdness that makes the codecov app report other coverage data.

**Comment by dyladan** [comment]
> Under the hood wtr already uses mocha, so that migration is free. However, we also need to replace webpack with rollup/esbuild - that requires some more changes. I was unable to get assert to work, so I replaced it with chai.

Removing webpack is something we've wanted to do anyway. In the core repo we've been working on updating tooling that was held back by the necessity to support old node versions which are being dropped soon.

> Coverage is collected, but I'm unsure if CI will correctly pick it up (EDIT: See later comments - the coverage is picked up, but something's weird with the merged report on codecov)

Coverage has been in a weird state in this repo for a while. It is likely not your fault.

**Comment by SimenB** [comment]
Cool! In that case, should I remove karma from the other packages as well?

**Review by svrnm — APPROVED** [review]
This will unblock the browser extension:-) , thank you @SimenB

File: plugins/web/opentelemetry-instrumentation-document-load/test/documentLoad.test.ts
Comment (pichlermarc) [reply plugins/web/opentelemetry-instrumentation-document-load/test/documentLoad.test.ts:39]: We used the webpack node polyfills for assert (see https://github.com/open-telemetry/opentelemetry-js-contrib/blob/434cab2539f35e00c4788cb48953b622950122dc/webpack.node-polyfills.js#L3-L6).

IMO not using node polyfills for tests that are only intended for the browser is a good idea - I'd be fine for this to switch to `chai`, would like to get an opinion on it by @martinkuba or @pkanal first though as they're the subject matter experts 🙂

**Comment by pichlermarc** [comment]
> Cool! In that case, should I remove karma from the other packages as well so we don't have to run both systems?
> 
> This PR can also be merged on its own, and then I can follow up for the other packages. Whatever's most convenient to you 👍

I think having follow ups to adapt the other instrumentations would be appreciated (once this PR has merged). We'd like to keep tooling in sync as much as possible. 🙂

**Comment by SimenB** [comment]
Yeah, happy to do follow-ups - should be quick. I was more asking if you want one big bang in a single PR or split up.

**Comment by SimenB** [comment]
Sweet 👍 will do once this is merged

**Comment by pichlermarc** [comment]
cc @martinkuba @pkanal (component-owners) - any objections? :slightly_smiling_face:
If not then I'd merge this in. :slightly_smiling_face:

**Review by pichlermarc — APPROVED** [review]
Brought this up in the SIG Meeting this week and there are no objections to switching. :slightly_smiling_face:

**Comment by pichlermarc** [comment]
@SimenB would you mind giving that PR an update? :slightly_smiling_face: I had been planning to merge it in a while ago but due to conflicts merging was blocked.

**Comment by SimenB** [comment]
image

hopefully CI is happy 🙂

**Comment by pichlermarc** [comment]
awesome, thanks :slightly_smiling_face:

**Pull request merged**
